Error Display (std::error::Error::fmt_error)

ijackson opened the pull request:
[Rendered](https://github.com/ijackson/rust-rfcs/blob/error/text/0000-error-display.md)

shepmaster commented:
I haven't fully read the RFC, but wanted to drop some additional prior art. [`snafu::report`](https://docs.rs/snafu/latest/snafu/attr.report.html) recognizes that many errors in the ecosystem basically implement `Display` as `write!(f, "My unique error text: {source}")`, which looks terrible when trying to format a nested error as you end up with:

```
1: C: B: A
2: B: A
3: A
```

`Report` post-processes those error messages to remove the duplicated content, producing:

```
1: C
2: B
3: A
```

joshtriplett commented:
cc @yaahc, whose `eyre` crate has an extensive reporting mechanism.

Stargateur commented:
I like the concept but I don't like much the limitation of current Display and Debug trait, you can't really do any sort of user configuration, like for example, carry information like if NO_COLOR is set in the environnement (you need to use static) thus I got no idea how we could solve this.

The fmt argument I mostly ignore all its features and just use it cause it has the write item.

I asked something very similar to tracing crate https://github.com/tokio-rs/tracing/issues/2226

Stargateur commented on text/0000-error-display.md:
```diff
@@ -0,0 +1,330 @@
+- Feature Name: separate_error_fmt
+- Start Date: 2027-07-19
```
wrong date

ijackson replied:
Oops, thanks :-).  Fixed.

ijackson commented on text/0000-error-display.md:
```diff
@@ -0,0 +1,330 @@
+/// each error displayed, and suppresses printing of the source if the
+/// source error text is textually contained within the previous error
+/// text.
+default impl<E> Display for E where E: Error + ErrorDisplay { ... }
```
This ought to have been deleted from here during editing, yes.  I've removed it.

ijackson commented:
Thanks to everyone for the review and comments.  I've updated the RFC.  Some more detailed reponses:

@shepmaster
> I haven't fully read the RFC, but wanted to drop some additional prior art. [`snafu::report`](https://docs.rs/snafu/latest/snafu/attr.report.html) [addresses this problem]

Thanks for the reference, indeed.  I wasn't aware of that.  I think that this textual deduplication approach has been independently invented (at least) twice shows it's a viable strategy.

@Stargateur
> I like the concept but I don't like much the limitation of current Display and Debug trait, you can't really do any sort of user configuration, like for example, carry information like if NO_COLOR is set in the environnement (you need to use static) thus I got no idea how we could solve this.

Indeed, this RFC won't eliminate the desire to do that, and it doesn't help with it very much.  For example, if you want coloured error messages, you'll still need to write a reporter that prints errors in colour, and arrange to call it everywhere you print an error.

But it does make it marginally easier: at least, with this RFC, you will (eventually) be able to reliably format the individual errors without their sources being mixed in.

@oxalica
> I doubt that if this can be achieved by the alternative flag `{:#}`

I have added a new section about this to Alternatives.  As you will see from my text, I think there are serious downsides to that approach.

@joshtriplett
> cc @yaahc, whose `eyre` crate has an extensive reporting mechanism.

Yes.  One way of looking at `eyre` is provides errors with a better  `Display`,  by wrapping them up.  I have added a reference.  `eyre` makes printing sources depend on `{}` vs `{:?}`, so I mention that too.

tgross35 commented:
GH isn't highlighting the code, could you change the fences to ` ```rust`?

kpreid commented:
I am glad to hear of a plan to make it easy and automatic to show `source` chains, thus relieving the tension between showing sources in `Display` (DWIM for printing and logging; makes it harder for error handling code to neglect to show the sources) and not showing sources in `Display` (enables clean structured chain printing). I don't like the result that nearly every error will automatically have the same default `Display` implementation, and that that implementation will have to _choose_ some chain formatting (the standard library mostly does not put _choices of syntax_ into `Display`, except for things like integers being base 10). But, an application can always introduce its own top-level error type with choice of `Display`, which is already common practice with e.g. `anyhow`.
